# Vogel's Approximation Method (VAM)/ Vogel'in Yaklaşım Yöntemi [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/faridnec/vogels-approx-method/blob/farid-main/vogels_am.ipynb)

*Authors : Muhammad Najmuddin Farid, Tea Skhurti, Mehmet Akif Kaya*

`EN` Vogel's Approximation Method (VAM) is an algorithm used for solving transportation problems in linear programming. A transportation problem involves minimizing the cost of transporting goods from multiple suppliers to multiple consumers. Vogel's method provides an iterative approach to find an initial feasible solution, which can then be improved using other optimization techniques.

`TR` Vogel'in Yaklaşım Yöntemi (VAM), doğrusal programlamada ulaştırma problemlerini çözmek için kullanılan bir algoritmadır. Ulaştırma problemi, malların birden fazla tedarikçiden birden fazla tüketiciye nakledilmesinin maliyetini en aza indirmeyi içerir. Vogel'in yöntemi, daha sonra diğer optimizasyon teknikleri kullanılarak geliştirilebilecek, başlangıçta uygun bir çözüm bulmak için yinelemeli bir yaklaşım sağlar.

**Reference**:  
BYJU's. (n.d.). Vogel's Approximation Method. BYJU's. https://byjus.com/maths/vogels-approximation-method/

### Tool(s)
Importing neccesary libraries to our project
- NumPy, a popular library for scientific computing
- relevant functions from `utils.py` and `vam.py` inside the `moduls` folder

In [1]:
from modules.utils import calculate_total_cost
from modules.vam import vogels_approximation_method
import numpy as np

## Scenario: Transportation Problem

`EN` You are the logistics manager for a company that produces goods at three different factories and needs to deliver them to four retail stores. The transportation costs (in dollars per unit) from each factory to each store are given in the following cost matrix:

`TR` Üç farklı fabrikada ürün üreten ve bunları dört perakende mağazasına teslim etmesi gereken bir şirketin lojistik müdürüsünüz. Her fabrikadan her mağazaya nakliye maliyetleri (birim başına dolar cinsinden) aşağıdaki maliyet matrisinde verilmiştir:

```markdown
                        ------------------------------------------------------------
                                  | Store 1 | Store 2 | Store 3 | Store 4 | Supply  
                        ------------------------------------------------------------
                        Factory 1 |   4     |   6     |   8     |   7     |   75    
                        Factory 2 |   2     |   6     |   8     |   5     |   125   
                        Factory 3 |   5     |   9     |   7     |   4     |   100  
                        ------------------------------------------------------------ 
                        Demand    |   80    |   65    |   70    |   85    |   300      
                        ------------------------------------------------------------
```
`EN` Using Vogel's Approximation Method (VAM) to find an initial feasible solution to minimize the total transportation cost.
1. Calculate penalties for each row and column.
2. Identify the row or column with the highest penalty.
3. Allocate units from the minimum cost cell in the selected row or column.
4. Adjust supply and demand.
5. Repeat steps 1-4 until all supply and demand requirements are satisfied.

`TR` Toplam taşıma maliyetini en aza indirecek başlangıç uygun çözümünü bulmak için Vogel'in Yaklaşım Yöntemini (VAM) kullanmak.
1. Her satır ve sütun için cezaları hesaplayın.
2. En yüksek cezaya sahip satır veya sütunu belirleyin.
3. Seçilen satır veya sütundaki minimum maliyet hücresinden birimleri tahsis edin.
4. Arz ve talebi ayarlayın.
5. Tüm arz ve talep gereksinimleri karşılanana kadar 1-4 arasındaki adımları tekrarlayın.

### Implementing supply, demand, and cost_matrix onto the code
Here, simply use numpy array to assign the value to be solved.

In [2]:
# Supply
supply = np.array([300, 400, 500])

# Demand
demand = np.array([250, 350, 400, 200])

# Cost Table/ Matrix
cost_matrix = np.array([
    [3, 1, 7, 4],
    [2, 6, 5, 9],
    [8, 3, 3, 2]
])

### Calculating penalties and choosing max penalty value
Using `vogels_approximation_method` to iterate and allocate optimized cost after calculating the penalties and choosing the maximum value both from row and column. Here is the visualized solution

![example2](./img/example2.png)

*Note :* In this particular example, notice on iteration 2. It does not allocate because the chosen maximun penalty value's cell is already fully allocated, thus it skips to the next iteration.

In [3]:
solution = vogels_approximation_method(supply, demand, cost_matrix)


Iteration 1:

Row Penalties: [2. 3. 1.]
Column Penalties: [1. 2. 2. 2.]
Max Penalty: 3.0
Allocating 250 units from Supplier 2 to Consumer 1

Iteration 2:

Row Penalties: [2. 1. 1.]
Column Penalties: [5. 2. 2. 2.]
Max Penalty: 5.0

Iteration 3:

Row Penalties: [3. 1. 1.]
Column Penalties: [0. 2. 2. 2.]
Max Penalty: 3.0
Allocating 300 units from Supplier 1 to Consumer 2

Iteration 4:

Row Penalties: [3. 1. 1.]
Column Penalties: [0. 3. 2. 2.]
Max Penalty: 3.0

Iteration 5:

Row Penalties: [0. 1. 1.]
Column Penalties: [0. 3. 2. 7.]
Max Penalty: 7.0
Allocating 200 units from Supplier 3 to Consumer 4

Iteration 6:

Row Penalties: [0. 1. 0.]
Column Penalties: [0. 3. 2. 0.]
Max Penalty: 3.0
Allocating 50 units from Supplier 3 to Consumer 2

Iteration 7:

Row Penalties: [0. 1. 5.]
Column Penalties: [0. 0. 2. 0.]
Max Penalty: 5.0
Allocating 250 units from Supplier 3 to Consumer 3

Iteration 8:

Row Penalties: [0. 1. 0.]
Column Penalties: [0. 0. 2. 0.]
Max Penalty: 2.0
Allocating 150 units from 

### Display Allocation Matrix
The optimized allocation for this specific scenario is as follows

In [4]:
print(solution)

[[  0 300   0   0]
 [250   0 150   0]
 [  0  50 250 200]]


### Calculate Minimum Cost
The minimum total transportation is calculated using`calculate_total_cost` and for this particular scenario
- cost =1×300+2×250+5×150+3×50+3×250+2×200= `2850`

In [5]:
calculate_total_cost(cost_matrix,solution)


Total Cost (Optimized): 2850 units

